content/courses/program-security/account-data-matching.md

```diff
@@ -304,114 +234,80 @@ pub struct InsecureWithdraw<'info> {
     pub token_program: Program<'info, Token>,
     pub authority: Signer<'info>,
 }
-
-#[account]
-pub struct Vault {
-    token_account: Pubkey,
-    authority: Pubkey,
-    withdraw_destination: Pubkey,
-}
 ```
 
-#### 2. Test `insecure_withdraw` instruction
+The critical flaw here is that while the instruction requires an `authority`
+signer, it doesn't verify whether the `authority` account passed to the
+instruction matches the `authority` stored in the `Vault` account. This means
+that any signer could potentially withdraw funds from the vault, even if they're
+not the authorized user.
 
-To prove that this is a problem, let’s write a test where an account other than
-the vault’s `authority` tries to withdraw from the vault.
+### 3. Test the Insecure Withdraw Instruction
 
-The test file includes the code to invoke the `initialize_vault` instruction
-using the provider wallet as the `authority` and then mints 100 tokens to the
-`vault` token account.
-
-Add a test to invoke the `insecure_withdraw` instruction. Use
-`withdrawDestinationFake` as the `withdrawDestination` account and `walletFake`
-as the `authority`. Then send the transaction using `walletFake`.
-
-Since there are no checks the verify the `authority` account passed into the
-instruction matches the values stored on the `vault` account initialized in the
-first test, the instruction will process successfully and the tokens will be
-transferred to the `withdrawDestinationFake` account.
+Let's write a test to demonstrate this vulnerability. Add the following test to
+your `tests/account-data-matching.ts` file:
 
 ```typescript
-describe("account-data-matching", () => {
-  ...
-  it("Insecure withdraw", async () => {
-    const tx = await program.methods
-      .insecureWithdraw()
-      .accounts({
-        vault: vaultPDA,
-        tokenAccount: tokenPDA,
-        withdrawDestination: withdrawDestinationFake,
-        authority: walletFake.publicKey,
-      })
-      .transaction()
-
-    await anchor.web3.sendAndConfirmTransaction(connection, tx, [walletFake])
-
-    const balance = await connection.getTokenAccountBalance(tokenPDA)
-    expect(balance.value.uiAmount).to.eq(0)
-  })
-})
+it("Insecure withdraw allows unauthorized access", async () => {
```

Comment:
This is very minor, but you're combining TSS/BDD style in your tests. It's super common everyone does it. `it` is supposed to read a little like a sentence, where `it` is the thing in the `describe` block, that's why it displays that way on the command line.

You should either make this `test("Insecure withdraw allows unauthorized access" ... (probably simplest) ` or `describe('insecure withdraw)` ....`it("allows unauthorized access" ... ` 

Do that for all the tests.